starlette/requests.py

```diff
@@ -19,7 +19,8 @@ class ClientDisconnect(Exception):
 
 
 class State:
-    pass
+    def __init__(self, state_dict={}):
+        self.__dict__.update(state_dict)
```

Comment:
This isn't right - you'll instead want to save a private `_state_dict` attribute on the instance, and read/modify it in `__getattr__`/`__setattr__`/`__delattr__`.

Reply:
When I did that previously, I got infinite recursive calls. I will debug again later to figure out why.

That said, a little off-tangent here...

I found out that I can pass around singleton objects by just assigning them to app object. Example:

app.settings[‘conf’] = conf

Then I can call it with:

request.app.settings[‘conf’].

This is similar to how Tornado does it.

Reply:
```
class State(object):
    def __init__(self, state={}):
        self._state = state

    def __setattr__(self, key, item: typing.Any) -> None:
        self._state[key] = item

    def __getattr__(self, key) -> typing.Any:
        return self._state[key]

    def __delattr__(self, key) -> None:
        del self._state[key]
```

## Error
```
starlette/requests.py:23: in __init__
    self._state = state
starlette/requests.py:26: in __setattr__
    self._state[key] = item
starlette/requests.py:29: in __getattr__
    return self._state[key]
starlette/requests.py:29: in __getattr__
    return self._state[key]
E   RecursionError: maximum recursion depth exceeded while calling a Python object
```

Reply:
I got it working using private `dict`:

```
class State(object):
    def __init__(self, state={}):
        self._state = state

    def __setattr__(self, key, value: typing.Any) -> None:
        if key == '_state':
            super(State, self).__setattr__(key, value)
        else:
            self._state[key] = value

    def __getattr__(self, key) -> typing.Any:
        if key == '_state':
            super(State, self).__getattr__(key)
        else:
            if key in self._state:
                return self._state[key]
            else:
                raise AttributeError("'{}' object has no attribute '{}'".format(self.__class__.__name__, key))

    def __delattr__(self, key) -> None:
        if key == '_state':
            super(State, self).__delattr__(key)
        else:
            del self._state[key]
```

But I am not sure if this code is more elegant than simply using default object's `__dict__` structure.

Reply:
Code coverage still stubbornly skips this line: `super(State, self).__getattr__(key)`

Even though I wrote several asserts to test the existence of keys:

```
def test_request_state_object():
    scope = {'state': {'old': 'foo'}}

    s = State(scope['state'])
    assert s._state == scope['state']
    assert getattr(s, '_state') == scope['state']

    s.new = 'value'                   # test __setattr__
    assert s.new == 'value'           # test __getattr__
    assert s._state['new'] == 'value' # test if inner _state dict is updated.

    del s.new # test __delattr__

    try:
        assert s.new == 'value' # will bombed with AttributeError
    except AttributeError as e:
        assert str(e) == "'State' object has no attribute 'new'"

    del s._state # Deleting _state dict should not bombed.
```